### DISCUSSION TITLE
Annotated dependency not treated as a default parameter

### DISCUSSION THREAD

**Discussion by karolpawlowski** [fastapi/fastapi#9393] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Iterator, Union

from fastapi import Depends, FastAPI
from sqlalchemy.orm.session import Session
from typing_extensions import Annotated

app = FastAPI()

def get_db_session() -> Iterator[Session]:
    """
    Meta database dependency for injection.
    """
    session: Session = Session()  # should be session maker, but leave it as is just for example

    try:
        yield session
    finally:
        session.close()


DBSession = Annotated[Session, Depends(get_db_session)]

@app.get("/annotated")
def get_items_annotated(
    enabled: Union[bool, None] = None,
    session: DBSession,  # marked as "non-default parameter follows default parameter!"
):
    return

@app.get("/old_way")
def get_items_old_way(
    enabled: Union[bool, None] = None,
    session: Session = Depends(get_db_session),  # works just fine
):
    return
```

### Description

When trying to migrate to annotated dependencies, they are marked as "non-default parameter follows default parameter!" when they follow parameters with default values. Isn't Depends also a kind of default value? An old approach works just fine.

Is changing parameter sequence the only way to fix this? Or is there another problem with annotated dependencies I found that wasn't fixed with 0.95.1 version?

Thanks in advance 🙏

### Operating System

macOS

### FastAPI Version

0.95.1

### Python Version

3.8

**Comment by nymous** [comment, marked as answer]
Hello!
This is not a FastAPI issue but a Python issue. Syntactically, you cannot define arguments without a default value after ones with a default value... except if you make all arguments keyword arguments.
This is documented in FastAPI docs: https://fastapi.tiangolo.com/tutorial/path-params-numeric-validations/#order-the-parameters-as-you-need-tricks
In your previous code all your parameters had a "default" value it worked. In the new code, you would change your endpoint to
```py
@app.get("/annotated")
def get_items_annotated(
    *,
    enabled: Union[bool, None] = None,
    session: DBSession,
):
    return
```

I think another way would be to also use `Annotated` for the `enabled` parameter, maybe like this (untested though ^^'):
```py
@app.get("/annotated")
def get_items_annotated(
    enabled: Annotated[Union[bool, None], Query(None)],
    session: DBSession,
):
    return

**Reply by karolpawlowski** [reply]
Thanks. I thought that maybe `Depends` could somehow fulfil this Python requirement under the hood and treat it as a keyword argument by default.
